### PR TITLE
First public release of promise type scl_docker_compose

### PR DESCRIPTION
**Pull request by basvandervlies**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @olehermanse can review this?

**Comment by olehermanse** [comment]
@basvandervlies Thank you for submitting this. Is there a reason to put `scl` in the name here? Does it rely on the SCL or is it otherwise very custom to something you are doing? If not, I'd suggest dropping this part in both the module name and promise type name.

File: cfbs.json
Comment (olehermanse) [inline cfbs.json:833]:
```suggestion
    "promise-type-docker-compose": {
```

File: cfbs.json
Comment (olehermanse) [inline cfbs.json:835]:
```suggestion
      "tags": ["promise-type", "experimental"],
```

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: cfbs.json
Comment (basvandervlies) [reply cfbs.json:833]: I will drop it and use `docker_compose`

File: cfbs.json
Comment (basvandervlies) [reply cfbs.json:835]: The things I submit are always `experimental`. Is this always for non-cfengine deverlopers ;-)?

Comment (nickanderson) [reply cfbs.json:835]: :rofl:

File: cfbs.json
Comment (olehermanse) [reply cfbs.json:833]: Please use only dashes in the Build module name (and underscores in the promise type in policy).

File: cfbs.json
Comment (olehermanse) [reply cfbs.json:835]: In general, all things are experimental until we've tested / feel confident about making them supported :)

File: cfbs.json
Comment (olehermanse) [inline cfbs.json:792]: It does not install, docker compose, right?

```suggestion
      "description": "Promise type to manage containers using docker compose",
```

Comment (basvandervlies) [reply cfbs.json:792]: Right is does not install `docker compose`

**Review by nickanderson — APPROVED** [review]
I looked at <https://github.com/basvandervlies/scl_modules/tree/7dd0ef2e2dca0c8669405f60d4690e577c51e435/promise-types/docker_compose>

In `docker_compose.sh` I found that `jq` is used, but it's listed as a requirement in the README along with docker compose which was an obvious requirement.

I don't really have an opinion, but I wondered if `enable.cf` ought to include a bundle to promise the installation of the required dependencies and have that bundle added to the bundlesequence.

I don't think it's necessary now, but food for thought.

I saw nothing mischievous in `docker_compose.sh`.

For grins I followed the docker compose getting started guide <https://docs.docker.com/compose/gettingstarted/>

I manually integrated the `docker_compose` promise type and gave it a go. But, it did not seem to be a **great success** as I got some unexpected output, `info: docker_compose.sh:service:'redis' state:'exited' is different then requested:'running'` and I was unable to access the web service which works when I change to that directory and manually run `docker compose up`. I guess this is on my end. `docker compose up` from the directory works and i can access the webapp fine, `docker compose --file /tmp/composetest/docker-compose.yml up` on the other hand brings up the containers, but i cam unable to access the webservice.

```
root@precision-5570:/var/cfengine/inputs# cf-agent -KI
    info: docker_compose.sh:service:'redis' state:'exited' is different then requested:'running'
[+] Running 1/1
 ⠿ Container composetest-redis-1  Started 0.5ss
    info: docker_compose.sh:service:'web' state:'exited' is different then requested:'running'
[+] Running 1/1
 ⠿ Container composetest-web-1  Started 0.5ss
root@precision-5570:/var/cfengine/inputs#
```

Altering the state to `stop` emits the following output and seems to bring the containers down as expected:

```
root@precision-5570:/var/cfengine/inputs# vi services/main.cf
root@precision-5570:/var/cfengine/inputs# cf-agent -KI
    info: docker_compose.sh:service:'redis' state:'running' is different then requested:'exited'
[+] Running 1/1
 ⠿ Container composetest-redis-1  Stopped 0.2ss
    info: docker_compose.sh:service:'web' state:'running' is different then requested:'exited'
[+] Running 1/1
 ⠿ Container composetest-web-1  Stopped 10.4ss
root@precision-5570:/var/cfengine/inputs#
```

I know that the start and stop keywords are legacy but perhaps these keywords should align with the docker states (running, exited) tho exited also has a connotation of having run at some point, so perhaps absent or a different word would make sense.

I did many stop start cycles in and out of CFEngine, making no actual changes and at some point I thought I saw that it was actually working for me when started by CFEngine, but I was unable to reproduce such a state on subsequent re-tries.

At any rate, I don't see any reason to block the merging of this PR.

&#x2026;. I couldn't resist playing more, and it seems that docker compose up from my console vs CFEngine was bringing it up on different IPs. I eventually saw it from `docker compose logs` .

So, LGTM!

Would be nice to get a runnable example for people to play with: Here is a start in case someone want's to pick it up and run.

Setting up

```sh
exec 2>&1
mkdir -p /tmp/composetest
:
```

`/tmp/composetest/app.py`

```python
import time

import redis
from flask import Flask

app = Flask(__name__)
cache = redis.Redis(host='redis', port=6379)

def get_hit_count():
    retries = 5
    while True:
        try:
            return cache.incr('hits')
        except redis.exceptions.ConnectionError as exc:
            if retries == 0:
                raise exc
            retries -= 1
            time.sleep(0.5)

@app.route('/')
def hello():
    count = get_hit_count()
    return 'Hello World! I have been seen {} times.\n'.format(count)
```

`/tmp/composetest/requirements.txt`

```text
flask
redis
```

`/tmp/composetest/Dockerfile`

```text
# syntax=docker/dockerfile:1
FROM python:3.7-alpine
WORKDIR /code
ENV FLASK_APP=app.py
ENV FLASK_RUN_HOST=0.0.0.0
RUN apk add --no-cache gcc musl-dev linux-headers
COPY requirements.txt requirements.txt
RUN pip install -r requirements.txt
EXPOSE 5000
COPY . .
CMD ["flask", "run"]
```

`/tmp/composetest/docker-compose.yml`

```yaml
version: "3.9"
services:
  web:
    build: .
    ports:
      - "8000:5000"
  redis:
    image: "redis:alpine"
```

```
root@precision-5570:/tmp/composetest# docker compose --file /tmp/composetest/docker-compose.yml up
[+] Running 2/0
 ⠿ Container composetest-redis-1  Created                                                                                                                                                                                                                                             0.0s
 ⠿ Container composetest-web-1    Created                                                                                                                                                                                                                                             0.0s
Attaching to composetest-redis-1, composetest-web-1
composetest-redis-1  | 1:C 27 Feb 2023 19:43:05.468 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
composetest-redis-1  | 1:C 27 Feb 2023 19:43:05.468 # Redis version=7.0.8, bits=64, commit=00000000, modified=0, pid=1, just started
composetest-redis-1  | 1:C 27 Feb 2023 19:43:05.468 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.469 * monotonic clock: POSIX clock_gettime
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.470 * Running mode=standalone, port=6379.
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.470 # Server initialized
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.470 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. Being disabled, it can can also cause failures without low memory condition, see https://github.com/jemalloc/jemalloc/issues/1328. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.471 * Loading RDB produced by version 7.0.8
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.471 * RDB age 46 seconds
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.471 * RDB memory usage when created 0.82 Mb
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.471 * Done loading RDB, keys loaded: 1, keys expired: 0.
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.471 * DB loaded from disk: 0.000 seconds
composetest-redis-1  | 1:M 27 Feb 2023 19:43:05.471 * Ready to accept connections
composetest-web-1    |  * Serving Flask app 'app.py'
composetest-web-1    |  * Debug mode: off
composetest-web-1    | WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
composetest-web-1    |  * Running on all addresses (0.0.0.0)
composetest-web-1    |  * Running on http://127.0.0.1:5000
composetest-web-1    |  * Running on http://172.18.0.3:5000
composetest-web-1    | Press CTRL+C to quit
```

**Comment by basvandervlies** [comment]
@nickanderson did it worked for you?  If not then we have o fix the issues before merging! At our site it works as expected bu we are running:
 * debian 11
 * docker compose version v2.15.1

There are plenty of `docker compose` files, maybe a simple hello world would be nice that is public available, eg:
 * https://github.com/docker/awesome-compose/tree/master/flask-redis

**Comment by basvandervlies** [comment]
Thanks I read it and still had the impression that it did not worked as expected ;-)

**Comment by basvandervlies** [comment]
How many approvals must I have before it is merged or is it still tested?

**Comment by nickanderson** [comment]
You got the approvals but no one merged it until now.

**Pull request merged**
